sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -450,3 +451,76 @@ case class ScalaUDAF(
 
   override def nodeName: String = udaf.getClass.getSimpleName
 }
+
+/**
+ * The internal wrapper used to hook a [[UserDefinedImperativeAggregator]] `udia` in the
+ * internal aggregation code path.
+ */
+case class ScalaUDIA[T](
+    children: Seq[Expression],
+    udia: UserDefinedImperativeAggregator[T],
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+  extends TypedImperativeAggregate[T]
+  with NonSQLExpression
+  with UserDefinedExpression
+  with ImplicitCastInputTypes
+  with Logging {
+
+  def dataType: DataType = udia.resultType
+
+  val inputTypes: Seq[DataType] = udia.inputSchema.map(_.dataType)
+
+  def nullable: Boolean = true
+
+  override lazy val deterministic: Boolean = udia.deterministic
+
+  def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ScalaUDIA[T] =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ScalaUDIA[T] =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  private[this] lazy val childrenSchema: StructType = {
+    val inputFields = children.zipWithIndex.map {
+      case (child, index) =>
+        StructField(s"input$index", child.dataType, child.nullable, Metadata.empty)
+    }
+    StructType(inputFields)
+  }
+
+  private lazy val inputProjection = {
+    val inputAttributes = childrenSchema.toAttributes
+    log.debug(
+      s"Creating MutableProj: $children, inputSchema: $inputAttributes.")
+    MutableProjection.create(children, inputAttributes)
+  }
+
+  private[this] lazy val inputToScalaConverters: Any => Any =
+    CatalystTypeConverters.createToScalaConverter(childrenSchema)
+
+  def createAggregationBuffer(): T = udia.initial
+
+  def update(buffer: T, input: InternalRow): T = {
+    val inrow = inputToScalaConverters(inputProjection(input)).asInstanceOf[Row]
```

Comment:
doesn't this actually do serde?

Reply:
I'm not entirely sure what that conversion does. It appeared to be important for maintaining behavior equivalence with ScalaUDAF. If anyone with expertise in Catalyst has additional insight, I would be interested.

Reply:
To clarify, whatever ser/de is happening here (formatting the input row) is not what is causing the problem, it is that the UDAF also is doing ser/de on the aggregator data structure.

Reply:
You need to convert the internal row produced by the `inputProjection` to an 'external' Row which the UDAF understands. To answers Reynold's question, yes this does deserialize data. It might be better to use a RowEncoder here if we can.

Reply:
@hvanhovell  would that be like 
```scala
val codec = RowEncoder(inputSchema).resolveAndBind()
// ...
val inrow = codec.fromRow(input)
```

Reply:
I tried using `RowEncoder` as above. It works fine but I didn't see any performance improvement from it, FWIW

Reply:
I did some more benchmarking experiments with codec versus the current input row prep. So far I am still failing to measure an improvement that passes statistical significance. There may be some way to improve this but I'd propose working it as a separate issue/pr